core/sequencer/server.go

```diff
@@ -380,3 +410,44 @@ func (s *Server) applyMutations(domainID string, mutatorFunc mutator.Func,
 	glog.V(2).Infof("applyMutations applied %v mutations to %v leaves", len(msgs), len(leaves))
 	return ret, nil
 }
+
+// HighWatermarks returns the total count and the highest watermark for each log.
+// batchSize is a limit on the total number of items represented by the returned watermarks.
+// TODO(gbelvin): Block until a minBatchSize has been reached or a timeout has occurred.
+func (s *Server) HighWatermarks(ctx context.Context, domainID string, starts Watermarks,
+	maxBatch int32) (int32, Watermarks, error) {
+	watermarks := make(Watermarks)
+	var total int32
+
+	// Ensure that we do not lose track of watermarks, even if they are no longer in the
```

Comment:
Hm, this sounds dangerous to me. What if you add/delete shards as it goes? I think you should be able to do the work when some are missing from this list (just assume they stay on their current mark).

Reply:
This struct currently serves two purposes - this distinction wasn't obvious to me until I traced down a subtle bug. 

1.  **Recreating a Map Revision** 
`sources` defines the ranges of inputs that constitute a map. These inputs are enough to recreate each map revision independently. As a result, if a log stays on it's current watermark, it can be dropped from this list, which is aesthetically desirable. 
2.  **Defining the *next* batch** 
When the sequencer takes a look around to figure out if any logs have grown since the last map revision, it needs the high watermarks of *all* logs that were considered for processing in the last map revision, even if those logs did not move.   ie. the sequencer needs the correct value of `start` for each potential source log.